In [1]:
import pandas as pd



In [2]:
import numpy as np


In [3]:
import matplotlib.pyplot as plt

In [4]:
import seaborn as sns


In [5]:
pd.set_option('display.max_columns', None)

In [6]:
df = pd.read_csv("../data/raw/ev_sales.csv")

In [7]:
df.head()

,Year,Month_Name,Date,State,Vehicle_Class,Vehicle_Category,Vehicle_Type,EV_Sales_Quantity
0,2014.0,jan,1/1/2014,Andhra Pradesh,ADAPTED VEHICLE,Others,Others,0.0
1,2014.0,jan,1/1/2014,Andhra Pradesh,AGRICULTURAL TRACTOR,Others,Others,0.0
2,2014.0,jan,1/1/2014,Andhra Pradesh,AMBULANCE,Others,Others,0.0
3,2014.0,jan,1/1/2014,Andhra Pradesh,ARTICULATED VEHICLE,Others,Others,0.0
4,2014.0,jan,1/1/2014,Andhra Pradesh,BUS,Bus,Bus,0.0


In [8]:
df.tail()

,Year,Month_Name,Date,State,Vehicle_Class,Vehicle_Category,Vehicle_Type,EV_Sales_Quantity
96840,2023.0,dec,12/1/2023,Andaman & Nicobar Island,MOTOR CAR,4-Wheelers,4W_Personal,1.0
96841,2023.0,dec,12/1/2023,Andaman & Nicobar Island,MOTOR CYCLE/SCOOTER-USED FOR HIRE,2-Wheelers,2W_Shared,5.0
96842,2023.0,dec,12/1/2023,Andaman & Nicobar Island,OMNI BUS,Bus,Bus,0.0
96843,2023.0,dec,12/1/2023,Andaman & Nicobar Island,THREE WHEELER (GOODS),3-Wheelers,3W_Goods,0.0
96844,2023.0,dec,12/1/2023,Andaman & Nicobar Island,THREE WHEELER (PASSENGER),3-Wheelers,3W_Shared,0.0


In [9]:
print(df.shape)

(96845, 8)


In [10]:
df.columns

Index(['Year', 'Month_Name', 'Date', 'State', 'Vehicle_Class',
       'Vehicle_Category', 'Vehicle_Type', 'EV_Sales_Quantity'],
      dtype='str')

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 96845 entries, 0 to 96844
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Year               96845 non-null  float64
 1   Month_Name         96845 non-null  str    
 2   Date               96845 non-null  str    
 3   State              96845 non-null  str    
 4   Vehicle_Class      96845 non-null  str    
 5   Vehicle_Category   96845 non-null  str    
 6   Vehicle_Type       96845 non-null  str    
 7   EV_Sales_Quantity  96845 non-null  float64
dtypes: float64(2), str(6)
memory usage: 10.8 MB


In [12]:
df.describe()

,Year,EV_Sales_Quantity
count,96845.000000,96845.000000
mean,2018.622768,37.108896
std,2.895581,431.566675
min,2014.000000,0.000000
25%,2016.000000,0.000000
50%,2019.000000,0.000000
75%,2021.000000,0.000000
max,2024.000000,20584.000000


In [13]:
df.duplicated().sum()

np.int64(0)

In [14]:
df["State"].nunique()

34

In [15]:
df["Vehicle_Category"].unique()

<ArrowStringArray>
['Others', 'Bus', '2-Wheelers', '4-Wheelers', '3-Wheelers']
Length: 5, dtype: str

In [16]:
df["Vehicle_Type"].unique()

<ArrowStringArray>
[            'Others',                'Bus',    'Institution Bus',
        '2W_Personal',          '4W_Shared',        '4W_Personal',
          '3W_Shared',        '3W_Personal',           '3W_Goods',
          '2W_Shared', '3W_Shared_LowSpeed',  '3W_Goods_LowSpeed']
Length: 12, dtype: str

In [17]:
df["Year"].unique()

array([2014., 2015., 2016., 2017., 2018., 2019., 2020., 2021., 2022.,
       2023., 2024.])

In [18]:
df["EV_Sales_Quantity"].sum()

np.float64(3593811.0)

# Dataset Understanding

- Dataset contains 96,845 records.
- No missing values are present.
- No duplicate records are found.
- Data covers EV sales from 2014 to 2024.
- Dataset contains 34 states.
- There are 5 vehicle categories and 12 vehicle types.
- EV sales data is highly skewed with many zero values.
- Data type optimization is required before analysis.

In [19]:
df_clean=df.copy()

In [26]:
df_clean.dtypes

Year                          int64
Month_Name                      str
Date                 datetime64[us]
State                           str
Vehicle_Class                   str
Vehicle_Category                str
Vehicle_Type                    str
EV_Sales_Quantity             int64
dtype: object

In [21]:
df_clean["Year"] = df_clean["Year"].astype(int)

In [25]:
df_clean["Date"] = pd.to_datetime(df_clean["Date"])

In [23]:
df_clean["EV_Sales_Quantity"]=df_clean["EV_Sales_Quantity"].astype(int)

In [28]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 96845 entries, 0 to 96844
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Year               96845 non-null  int64         
 1   Month_Name         96845 non-null  str           
 2   Date               96845 non-null  datetime64[us]
 3   State              96845 non-null  str           
 4   Vehicle_Class      96845 non-null  str           
 5   Vehicle_Category   96845 non-null  str           
 6   Vehicle_Type       96845 non-null  str           
 7   EV_Sales_Quantity  96845 non-null  int64         
dtypes: datetime64[us](1), int64(2), str(5)
memory usage: 10.0 MB


In [29]:
df_clean["Month_Number"] = df_clean["Date"].dt.month

In [30]:
df_clean["Quarter"] = df_clean["Date"].dt.quarter

In [31]:
df_clean.head()

,Year,Month_Name,Date,State,Vehicle_Class,Vehicle_Category,Vehicle_Type,EV_Sales_Quantity,Month_Number,Quarter
0,2014,jan,2014-01-01,Andhra Pradesh,ADAPTED VEHICLE,Others,Others,0,1,1
1,2014,jan,2014-01-01,Andhra Pradesh,AGRICULTURAL TRACTOR,Others,Others,0,1,1
2,2014,jan,2014-01-01,Andhra Pradesh,AMBULANCE,Others,Others,0,1,1
3,2014,jan,2014-01-01,Andhra Pradesh,ARTICULATED VEHICLE,Others,Others,0,1,1
4,2014,jan,2014-01-01,Andhra Pradesh,BUS,Bus,Bus,0,1,1


In [36]:
df_clean.to_csv(
    "../data/proceed/cleaned_ev_sales.csv",
    index=False
)

## Operations Performed

- Created a copy of the original dataset.
- Converted Year from float to integer.
- Converted EV Sales Quantity from float to integer.
- Converted Date to datetime format.
- Created Month_Number feature.
- Created Quarter feature.
- Saved the cleaned dataset.